File: py800mon/monitor/displaylist.py
from ..actions import Actions
from ..app import VisualRpcComponent
from ..atari.displaylist import DLPTRS_ADDR, DMACTL_ADDR, DMACTL_HW_ADDR, decode_displaylist
from ..rpc import RpcException
from ..ui import Color, GridWidget
from .appstate import state


class DisplayListViewer(VisualRpcComponent):
    def __init__(self, rpc, window):
        super().__init__(rpc, window)
        self.grid = GridWidget(window)
        self.grid.add_column("address", width=5, attr=Color.ADDRESS.attr())
        self.grid.add_column("description", width=0, attr=Color.TEXT.attr())
        self._dmactl = 0

    async def update(self):
        try:
            start_addr = await self.rpc.read_vector(DLPTRS_ADDR)
            dump = await self.rpc.read_display_list()
            dmactl = await self.rpc.read_byte(DMACTL_ADDR)
            if (dmactl & 0x03) == 0:
                dmactl = await self.rpc.read_byte(DMACTL_HW_ADDR)
        except RpcException:
            return False
        else:
            dlist = decode_displaylist(start_addr, dump)
            self.app.dispatch_action(Actions.SET_DLIST, (dlist, dmactl))
            self._dmactl = dmactl
            return True

    def render(self, force_redraw=False):
        rows = []
        for count, entry in state.dlist.compacted_entries():
            if count > 1:
                desc = f"{count}x {entry.description}"
            else:
                desc = entry.description
            rows.append((f"{entry.addr:04X}:", desc))
        self.grid.set_data(rows)
        if rows and self.grid.selected_row is None:
            self.grid.set_selected_row(0)
        self.grid.render()

    def handle_input(self, ch):
        return self.grid.handle_input(ch)

File: go800mon/a800mon/monitor/displaylist.go
package monitor

import (
	"context"
	"fmt"

	. "go800mon/a800mon"
	atari "go800mon/a800mon/atari"
)

type DisplayListViewer struct {
	BaseWindowComponent
	rpc          *RpcClient
	lastSnapshot string
	grid         *GridWidget
}

func NewDisplayListViewer(rpc *RpcClient, window *Window) *DisplayListViewer {
	grid := NewGridWidget(window)
	grid.AddColumn("address", 5, ColorAddress.Attr(), nil)
	grid.AddColumn("description", 0, ColorText.Attr(), nil)
	return &DisplayListViewer{
		BaseWindowComponent: NewBaseWindowComponent(grid.Window()),
		rpc:                 rpc,
		grid:                grid,
	}
}

func (v *DisplayListViewer) Update(ctx context.Context) (bool, error) {
	startAddr, err := v.rpc.ReadVector(ctx, atari.DLPTRSAddr)
	if err != nil {
		return false, nil
	}
	dump, err := v.rpc.ReadDisplayList(ctx)
	if err != nil {
		return false, nil
	}
	dmactl, err := v.rpc.ReadByte(ctx, atari.DMACTLAddr)
	if err != nil {
		return false, nil
	}
	if (dmactl & 0x03) == 0 {
		if hw, err := v.rpc.ReadByte(ctx, atari.DMACTLHWAddr); err == nil {
			dmactl = hw
		}
	}
	dlist := atari.DecodeDisplayList(startAddr, dump)
	if app := v.App(); app != nil {
		app.DispatchAction(
			ActionSetDList,
			DListUpdate{DList: dlist, DMACTL: dmactl},
		)
	}
	snapshot := fmt.Sprintf("%04X|%02X|%d", startAddr, dmactl, len(dlist.Entries))
	if len(dlist.Entries) > 0 {
		first := dlist.Entries[0]
		last := dlist.Entries[len(dlist.Entries)-1]
		snapshot += fmt.Sprintf("|%02X-%04X|%02X-%04X", first.Command, first.Arg, last.Command, last.Arg)
	}
	if v.lastSnapshot == snapshot {
		return false, nil
	}
	v.lastSnapshot = snapshot
	return true, nil
}

func (v *DisplayListViewer) Render(_force bool) {
	st := State()
	g := v.grid
	rows := make([][]string, 0, len(st.DList.Entries))
	for _, c := range st.DList.Compacted() {
		addr := fmt.Sprintf("%04X:", c.Entry.Addr)
		desc := c.Entry.Description()
		if c.Count > 1 {
			desc = fmt.Sprintf("%dx %s", c.Count, c.Entry.Description())
		}
		rows = append(rows, []string{addr, desc})
	}
	g.SetData(rows)
	if len(rows) > 0 {
		if _, ok := g.SelectedRow(); !ok {
			idx := 0
			g.SetSelectedRow(&idx)
		}
	} else {
		g.SetSelectedRow(nil)
	}
	g.Render()
}

func (v *DisplayListViewer) HandleInput(ch int) bool {
	return v.grid.HandleInput(ch)
}
